### ISSUE TITLE
Add support for Levoit Vital 200s air purifier

### ISSUE THREAD

**Issue by MollyrH** [webdjoe/pyvesync#178]
When/if possible, could support for the new model Vital 200s air purifier be added?

Thank you in advance!

**Comment by webdjoe** [comment]
I would need packet captures. Capturing packets has become more difficult due to ssl pinning, the only current way is by following the directions here in the feature requests section of the Readme

**Comment by M1K3SD1** [comment]
@MollyrH I can try and help with the captures if you'd like. It's not easy at all and I'll need your cooperation if you want, but it can be done.

Since I have a packet capturing environment ready but no access to such device, I can suggest you creating a temporary account in vesync with a temporary email, connect your device to this account.

Send me these credentials privately so I can connect and capture the packets, after which I will log out and you can disconnect you device from this temp account.

Note that this will give me access to your actual device, so if you're not comfortable with this don't do this.

**Comment by kschlichter** [comment]
@M1K3SD1, I have one on a dedicated account you can use. I'm not sure how to get you the credentials though.  Email me at pyvesync178@gmail.com and I can send you the VeSync password.

**Comment by M1K3SD1** [comment]
@kschlichter I captured all functions, cleaned the file, and send it to @webdjoe.
I've logged off, you can now disconnect your device and use it.
Thanks!

**Comment by webdjoe** [comment]
@M1K3SD1 Thank you for the assist! @kschlichter Check out #188 I've added Vital 200S support to that branch. Please test it out

**Comment by kschlichter** [comment]
I must be doing something wrong.  I've entered the Home Assistant Docker container and replaced /usr/local/lib/python3.11/site-packages/pyvesync/vesyncfan.py with the raw code from [#181 188](https://github.com/webdjoe/pyvesync/pull/188) but when I call the service: vesync.update_devices, I get this in the logs:
`This error originated from a custom integration.

Logger: custom_components.vesync.common
Source: custom_components/vesync/common.py:78
Integration: VeSync
First occurred: 4:48:22 PM (1 occurrences)
Last logged: 4:48:22 PM

Unknown device type Vital200s LAP-V201S-WUS (enable debug for more info)`

What's the proper way to test these changes?

**Comment by webdjoe** [comment]
You are unable to test this in home assistant unless you clone the HA repo and replace the version requirements. Please see the https://github.com/webdjoe/pyvesync/tree/2.18#feature-requests for testing. I released in 2.18 on pypi so you can now just `pip install pyvesync` directly. Let me know if you come across any bugs

**Comment by kschlichter** [comment]
I'm not great at this but I don't think it's reading the status of my Vital 200s correctly.  Jump to the end for the results.  Maybe I'm doing something wrong, so I'm being verbose.
```
python_test$ python3 --version
Python 3.10.6

python3 -m venv pyvesync-venv

source pyvesync-venv/bin/activate

pip install pyvesync
Collecting pyvesync
  Using cached pyvesync-2.1.8-py2.py3-none-any.whl (50 kB)
Collecting requests>=2.20.0
  Using cached requests-2.31.0-py3-none-any.whl (62 kB)
Collecting charset-normalizer<4,>=2
  Using cached charset_normalizer-3.2.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (201 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2023.5.7-py3-none-any.whl (156 kB)
Collecting idna<4,>=2.5
  Using cached idna-3.4-py3-none-any.whl (61 kB)
Collecting urllib3<3,>=1.21.1
  Using cached urllib3-2.0.4-py3-none-any.whl (123 kB)
Installing collected packages: urllib3, idna, charset-normalizer, certifi, requests, pyvesync
Successfully installed certifi-2023.5.7 charset-normalizer-3.2.0 idna-3.4 pyvesync-2.1.8 requests-2.31.0 urllib3-2.0.4
```

The test.py script at the [2.18#feature-requests](https://github.com/webdjoe/pyvesync/tree/2.18#feature-requests) link had two if statements missing ':' and it couldn't see any devices.  Instead, I've done:

```
import sys
import logging
import json
from pyvesync import VeSync

logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)

manager = VeSync("<user>", "<password>", debug=True)
if manager.login() == False: # Added a ':'
  logger("Unable to login")
```
That can log me in, but gets no devices.  
```
>>> for dev in manager.fans:
...   dev
... 
>>> manager.fans
manager.fans
>>> manager.fans
[]
>>> 
```
Adding a call to manager.update() helps and now I get:
```
my_device_name = "Vital200s"
fan = None
for dev in manager.fans:
  dev
>>> my_device_name = "Vital200s"
>>> fan = None
>>> for dev in manager.fans:
...   dev
... 
DevClass: VeSyncAirBypass,                Name:Core300s Guest, Device No: None,                DevStatus: on, CID: vsaq52310844bfeaecb258c87e1a1704
DevClass: VeSyncAirBypass,                Name:Core300s Kathryn, Device No: None,                DevStatus: off, CID: vsaqb540e4d4b56a4ea7187cf33751cb
DevClass: VeSyncAirBypass,                Name:Core300s MBR, Device No: None,                DevStatus: off, CID: vsaqaed95f64e128462f5dd697ac14c9
DevClass: VeSyncVital,                Name:Vital200s, Device No: None,                DevStatus: off, CID: vsaq3699c8f4454ae631f2279e2f85c3

```
Huzzah.  I can see all of my devices.  I haven't figured out how to DO anything with that but it's notable that **the DevStatus is correct for the three Core300s devices (on / off / off) but Vital200s says DevStatus: off when it's on.**  If I can figure out how to test the other features I'll update some more.  For now, I'm unsure how to move forward.

**Comment by webdjoe** [comment]
Thanks for letting me know, I updated the README to include the `update()` method and the missing colons.

You're almost there! Just run the methods and test the properties included [Standard Air Purifier Properties and Methods](https://github.com/webdjoe/pyvesync#standard-air-purifier-properties--methods) and in the [Vital 100s/200s Properties and Methods](https://github.com/webdjoe/pyvesync#levoit-vital-100s200s-properties-and-methods)

Here is a script that I've tested (on all of the standard Air Purifier properties and methods).  Or you can debug and run VSCode interactively in the debug console with a breakpoint after the device is found.
```python
import logging
import json
from pyvesync import VeSync

logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)

DEVICE_NAME = "DEVICE NAME"

USERNAME = "USER NAME"
PASSWORD = "PASSWORD"

def test_device():
    # Instantiate VeSync class and login
    manager = VeSync(USERNAME, PASSWORD, debug=True)
    if manager.login() == False:
        logger.debug("Unable to login")
        return

    # Pull and update devices
    manager.update()

    fan = None
    logger.debug(str(manager.fans))

    for dev in manager.fans:
        # Print all device info
        logger.debug(dev.device_name + "\n")
        logger.debug(dev.display())

        # Find correct device
        if dev.device_name.lower() == DEVICE_NAME.lower():
            fan = dev
            break

    if fan == None:
        logger.debug("Device not found")
        logger.debug("Devices found - \n" + manager._dev_list)
        return


    logger.debug('--------------%s-----------------' % fan.device_name)
    logger.debug(dev.display())
    logger.debug(dev.displayJSON())
    # Test all device methods and functionality
    # Test Properties
    logger.debug("Fan is on - %s", fan.is_on)
    logger.debug("Modes - %s", fan.modes)
    logger.debug("Fan Level - %s", fan.fan_level)
    logger.debug("Fan Air Quality - %s", fan.air_quality)
    logger.debug("Screen Status - %s", fan.screen_status)

    fan.turn_on()
    fan.turn_off()
    fan.sleep_mode()
    fan.auto_mode()
    fan.manual_mode()
    fan.change_fan_speed(3)
    fan.change_fan_speed(2)
    fan.child_lock_on()
    fan.child_lock_off()
    fan.turn_off_display()
    fan.turn_on_display()

    fan.set_light_detection_on()
    logger.debug(fan.light_detection_state)
    logger.debug(fan.light_detection)

    # Only on Vital 200S
    fan.pet_mode()

    logger.debug("Set Fan Speed - %s", fan.set_fan_speed)
    logger.debug("Current Fan Level - %s", fan.fan_level)
    logger.debug("Current mode - %s", fan.mode)

    # Display all device info
    logger.debug(dev.display())
    logger.debug(dev.displayJSON())

if __name__ == "__main__":
    logger.debug("Testing device")
    test_device()
```

**Comment by kschlichter** [comment]
That was extremely helpful.  Thank you for all the work to get here.  I've found three bugs but _most_ of this works perfectly already.   I wish I could submit code changes here but I've had to limit my contributions to just being thorough.  

1. The `fan.is_on` boolean will always say `False`.  I think it's expecting the wrong dictionary key.  
If the purifier is off, `manager.update()` returns:
```
2023-07-23 16:39:02,282 - DEBUG - =======call_api=============================
2023-07-23 16:39:02,282 - DEBUG - [post] calling '/cloud/v2/deviceManaged/bypassV2' api
2023-07-23 16:39:02,282 - DEBUG - API call URL: 
  https://smartapi.vesync.com/cloud/v2/deviceManaged/bypassV2
2023-07-23 16:39:02,283 - DEBUG - API call headers: 
  {"Content-Type": "application/json; charset=UTF-8", "User-Agent": "okhttp/3.12.1"}
2023-07-23 16:39:02,283 - DEBUG - API call json: 
  {"timeZone": "America/New_York", "acceptLanguage": "en", "accountID": "##_REDACTED_##", "token": "##_REDACTED_##", "appVersion": "2.8.6", "phoneBrand": "SM N9005", "phoneOS": "Android", "traceId": "1690155542", "method": "bypassV2", "debugMode": false, "deviceRegion": "US", "cid": "##_REDACTED_##", "deviceId": "vsaq3699c8f4454ae631f2279e2f85c3", "configModule": "VS_WFON_APR_LAP-V201S-WUS_US", "configModel": "VS_WFON_APR_LAP-V201S-WUS_US", "payload": {"method": "getPurifierStatus", "source": "APP", "data": {}}}
2023-07-23 16:39:02,889 - DEBUG - API response: 

  {"traceId": "1690155542", "code": 0, "msg": "request success", "module": null, "stacktrace": null, "result": {"traceId": "1690155542", "code": 0, "result": {"powerSwitch": 0, "filterLifePercent": 55, "workMode": "auto", "manualSpeedLevel": 3, "fanSpeedLevel": 255, "AQLevel": 1, "PM25": 1, "screenState": 0, "childLockSwitch": 0, "screenSwitch": 1, "lightDetectionSwitch": 0, "environmentLightState": 0, "autoPreference": {"autoPreferenceType": "default", "roomSize": 0}, "scheduleCount": 0, "timerRemain": 0, "efficientModeTimeRemain": 0, "sleepPreference": {"sleepPreferenceType": "default", "cleaningBeforeBedSwitch": 1, "cleaningBeforeBedSpeedLevel": 3, "cleaningBeforeBedMinutes": 5, "whiteNoiseSleepAidSwitch": 1, "whiteNoiseSleepAidSpeedLevel": 1, "whiteNoiseSleepAidMinutes": 45, "duringSleepSpeedLevel": 5, "duringSleepMinutes": 480, "afterWakeUpPowerSwitch": 1, "afterWakeUpWorkMode": "auto", "afterWakeUpFanSpeedLevel": 1}, "errorCode": 0}}} 
 
2023-07-23 16:39:02,891 - DEBUG - No configuration found in purifier status
```
…and if the fan is _on_ I get:
```
2023-07-23 16:41:52,289 - DEBUG - =======call_api=============================
2023-07-23 16:41:52,290 - DEBUG - [post] calling '/cloud/v2/deviceManaged/bypassV2' api
2023-07-23 16:41:52,290 - DEBUG - API call URL: 
  https://smartapi.vesync.com/cloud/v2/deviceManaged/bypassV2
2023-07-23 16:41:52,291 - DEBUG - API call headers: 
  {"Content-Type": "application/json; charset=UTF-8", "User-Agent": "okhttp/3.12.1"}
2023-07-23 16:41:52,291 - DEBUG - API call json: 
  {"timeZone": "America/New_York", "acceptLanguage": "en", "accountID": "##_REDACTED_##", "token": "##_REDACTED_##", "appVersion": "2.8.6", "phoneBrand": "SM N9005", "phoneOS": "Android", "traceId": "1690155712", "method": "bypassV2", "debugMode": false, "deviceRegion": "US", "cid": "##_REDACTED_##", "deviceId": "vsaq3699c8f4454ae631f2279e2f85c3", "configModule": "VS_WFON_APR_LAP-V201S-WUS_US", "configModel": "VS_WFON_APR_LAP-V201S-WUS_US", "payload": {"method": "getPurifierStatus", "source": "APP", "data": {}}}
2023-07-23 16:41:52,891 - DEBUG - API response: 

  {"traceId": "1690155712", "code": 0, "msg": "request success", "module": null, "stacktrace": null, "result": {"traceId": "1690155712", "code": 0, "result": {"powerSwitch": 1, "filterLifePercent": 55, "workMode": "auto", "manualSpeedLevel": 3, "fanSpeedLevel": 0, "AQLevel": 1, "PM25": 1, "screenState": 1, "childLockSwitch": 0, "screenSwitch": 1, "lightDetectionSwitch": 0, "environmentLightState": 0, "autoPreference": {"autoPreferenceType": "default", "roomSize": 0}, "scheduleCount": 0, "timerRemain": 0, "efficientModeTimeRemain": 0, "sleepPreference": {"sleepPreferenceType": "default", "cleaningBeforeBedSwitch": 1, "cleaningBeforeBedSpeedLevel": 3, "cleaningBeforeBedMinutes": 5, "whiteNoiseSleepAidSwitch": 1, "whiteNoiseSleepAidSpeedLevel": 1, "whiteNoiseSleepAidMinutes": 45, "duringSleepSpeedLevel": 5, "duringSleepMinutes": 480, "afterWakeUpPowerSwitch": 1, "afterWakeUpWorkMode": "auto", "afterWakeUpFanSpeedLevel": 1}, "errorCode": 0}}} 
 
2023-07-23 16:41:52,893 - DEBUG - No configuration found in purifier status
```
So `"powerSwitch": 0` changes to `"powerSwitch": 1`, when the purifier turns on while my Core 300 purifiers have `"enabled": false` or `"enabled": true` tracking the power status in that dictionary, depending on their state.  

2. The `fan.set_light_detection_on()` and `fan.set_light_detection_off()` functions fail.  If the display is off and I try to turn it on:
```
>>> fan.set_light_detection_on()
2023-07-23 16:46:50,951 - DEBUG - =======call_api=============================
2023-07-23 16:46:50,951 - DEBUG - [post] calling '/cloud/v2/deviceManaged/bypassV2' api
2023-07-23 16:46:50,951 - DEBUG - API call URL: 
  https://smartapi.vesync.com/cloud/v2/deviceManaged/bypassV2
2023-07-23 16:46:50,951 - DEBUG - API call headers: 
  {"Content-Type": "application/json; charset=UTF-8", "User-Agent": "okhttp/3.12.1"}
2023-07-23 16:46:50,952 - DEBUG - API call json: 
  {"timeZone": "America/New_York", "acceptLanguage": "en", "accountID": "##_REDACTED_##", "token": "##_REDACTED_##", "appVersion": "2.8.6", "phoneBrand": "SM N9005", "phoneOS": "Android", "traceId": "1690156010", "method": "bypassV2", "debugMode": false, "deviceRegion": "US", "cid": "##_REDACTED_##", "deviceId": "vsaq3699c8f4454ae631f2279e2f85c3", "configModule": "VS_WFON_APR_LAP-V201S-WUS_US", "configModel": "VS_WFON_APR_LAP-V201S-WUS_US", "payload": {"method": "setLightDetectionSwitch", "source": "APP", "data": {"lightDetectionSwitch": 1}}}
2023-07-23 16:46:51,585 - DEBUG - API response: 

  {"traceId": "1690156010", "code": 0, "msg": "request success", "module": null, "stacktrace": null, "result": {"traceId": "1690156010", "code": -1}} 
 
True
```
…and after turning it on through the app, then waiting the 30 seconds so manager.update() will run:
```
>>> fan.set_light_detection_off()
2023-07-23 16:49:49,096 - DEBUG - =======call_api=============================
2023-07-23 16:49:49,096 - DEBUG - [post] calling '/cloud/v2/deviceManaged/bypassV2' api
2023-07-23 16:49:49,096 - DEBUG - API call URL: 
  https://smartapi.vesync.com/cloud/v2/deviceManaged/bypassV2
2023-07-23 16:49:49,097 - DEBUG - API call headers: 
  {"Content-Type": "application/json; charset=UTF-8", "User-Agent": "okhttp/3.12.1"}
2023-07-23 16:49:49,097 - DEBUG - API call json: 
  {"timeZone": "America/New_York", "acceptLanguage": "en", "accountID": "##_REDACTED_##", "token": "##_REDACTED_##", "appVersion": "2.8.6", "phoneBrand": "SM N9005", "phoneOS": "Android", "traceId": "1690156189", "method": "bypassV2", "debugMode": false, "deviceRegion": "US", "cid": "##_REDACTED_##", "deviceId": "vsaq3699c8f4454ae631f2279e2f85c3", "configModule": "VS_WFON_APR_LAP-V201S-WUS_US", "configModel": "VS_WFON_APR_LAP-V201S-WUS_US", "payload": {"method": "setLightDetectionSwitch", "source": "APP", "data": {"lightDetectionSwitch": 0}}}
2023-07-23 16:49:49,742 - DEBUG - API response: 

  {"traceId": "1690156189", "code": 0, "msg": "request success", "module": null, "stacktrace": null, "result": {"traceId": "1690156189", "code": -1}} 
 
True
```
The result code is -1 for both commands, while others that succeed result in code 0.  Interestingly though, If light detection is already on, I get:
```
>>> fan.set_light_detection_on()
2023-07-23 16:52:12,625 - DEBUG - Light Detection is already set to 1
True
```

3. The `Online` status is always reported as `None`.  This status is reported correctly for my Core 300S purifiers though.
```
>>> dev.display()
Device Name:.................. Vital200s
Model: ....................... LAP-V201S-WUS
Subdevice No: ................ None
Status: ...................... off
Online: ...................... None
```

Finally, `dev.display()` doesn't have the status of the light detection option, or whether it's detecting light, so I can't tell if this will be something Home Assistant will be able to report as part of the status.  This isn't a bug, I suppose; it's a feature request to add `Light Detection On: True/False/NA` and `Light Detected: True/False/NA` to the display() function (where NA is the status for fans that don't support the feature).

I've played around with these functions quite a bit, and all work perfectly:
```
fan.turn_on()
fan.turn_off()
fan.sleep_mode()
fan.auto_mode()
fan.manual_mode()
fan.change_fan_speed(3)
fan.change_fan_speed(2)
fan.child_lock_on()
fan.child_lock_off()
fan.turn_off_display()
fan.turn_on_display()
fan.pet_mode()

>>> dev.display()
Device Name:.................. Vital200s
Model: ....................... LAP-V201S-WUS
Subdevice No: ................ None
Status: ...................... off                   # Bug reported above
Online: ...................... None              # Bug reported above
Type: ........................ wifi-air
CID: ......................... vsaq3699c8f4454ae631f2279e2f85c3
UUID: ........................ dadcffdc-5eb5-4e34-bce4-a1bf16e0980c
Mode: ........................ auto 
Filter Life: ................. 55 percent
Fan Level: ................... 0     # 0 is a fan speed less than 1, but not off.  It does seem to change, even when the status is 'auto'
Display: ..................... True 
Child Lock: .................. False 
Night Light: ................. off                 # Feature from a different purifier; not tested.
Display Config: .............. False         # I don't know what this is
Display_Forever Config: ...... False   # I don't know what this is
Air Quality Level: ........... 0                # I get "very good" in the app but numbers are easier to evaluate.  
Air Quality Value: ........... 1 ug/m3     # I can't see this in the app, and I'm so happy to see it here!  

```

Thanks again for all of the effort you've put into this.  What you have in 2.18 is functional and I appreciate it very much.  These last things would be nice but I'm sure you're busy.

**Comment by chipdarippa** [comment]
@M1K3SD1, is there anyway you could assist us with capturing data for the Levoit EverestAir over in issue #170? I can share my device with a throwaway email account and give that login information to you.

**Comment by dag81** [comment]
Thank you @kschlichter for posting the unit response above 👍

In regards to the response payload shown @webdjoe I suspect on this line:

https://github.com/webdjoe/pyvesync/blob/2d70bb4c12cdd2c64243aa62a4cd6b0548c99a04/src/pyvesync/vesyncfan.py#L836

air_quality is not in the response object from what I can see - so my guess would be this has been renamed to AQLevel instead of air_quality in the response's payload?

**Comment by obi11235** [comment]
I made a PR with some crude fixes @kschlichter called out. I couldn't address the light detection issue since I don't know what the api calls should look like. I'm testing it on a LAP-V201S-WUS.

https://github.com/webdjoe/pyvesync/pull/201

**Comment by webdjoe** [comment]
Fixed in #198 and released in 2.1.9. I updated `displayjson()` method but it needs a major overhaul and chances are it is not used by HA or any other libraries
